Spigot 1.13, Java 1.8

Jikoo opened the pull request:
* Fixed double chests/other containers
  * All multiblock structures (double chests, beds, etc.) will attempt to heal both blocks at once. If a single block fails, the whole heal fails.
  * Future Bisected additions will be considered multiblocks by default, but may need to be excluded the way stairs and slabs are to properly heal block-by-block.
* Added support for most new blocks
  * Missing coral and some other things - still have yet to revisit the concept of NeighborBlocks entirely, which should handle cases in a more generic fashion.
* Fixed a lot of potential future issues associated with redundant explosion replacement entries for "dependent" multiblocks - mostly ignoring more duplicate entries in CreeperExplosion
* Changed configs to use Material names. This WILL break existing configs using IDs. Not sure how well configs work at all, have yet to test. Not a priority for me, there are a dozen claim/anti-grief plugins out there.

Currently not ready to be pulled, input welcome.

TODO:
- [x] Split up PR into multiple commits
- [ ] Revisit Neighbor blocks
- [x] Fix formatting
- [ ] Configurations
  - [ ] Test, make sure materials work properly
  - [x] Investigate converter
- [ ] Remove local dependencies

I also did not bump the version, figured that'd be up to you to decide.

Jikoo committed:
Update to 1.13

(cherry picked from commit 2909731)

nitnelave commented:
Wow, thanks for the PR! I'll try to have a look this weekend :)

 In general, though, it's *much* easier to get PRs accepted if you decrease the scope: 5 small PRs get in faster than one huuuuge PR, and issues/bugs can be identified and addressed more easily.

That is especially true if you decide to make breaking changes ;)

nitnelave commented:
I had a very quick look, and it would really help if you did at least the following splits:

- Formatting corrections (trailing whitespace, line wrapping, etc...)
- Dependency changes (removing the lib/ folder, changing the pom.xml file)
- Ideally, each main feature as a separate change (multiblock, new blocks, redundancy, ...)
- Breaking changes: we have to consider each of them separately, and separately from the rest of the changes that are just clear improvements.

If you need help with that, I'll try my best to help!

Jikoo commented:
>* Formatting corrections (trailing whitespace, line wrapping, etc...)

Definitely on my list, the trailing whitespace removal was a mistake I thought I'd corrected. I also need to go back and redo the formatting to match your style.

>* Dependency changes (removing the lib/ folder, changing the pom.xml file)

If that's how you want to go, can do. Didn't want to mess with the setup you have.

>* Ideally, each main feature as a separate change (multiblock, new blocks, redundancy, ...)

Most multiblocks are tied to BlockData, new in 1.13. I can definitely move some of the simplifications to other methods into a separate commit.It should also be possible to split off the chest/doublechest section of it if you want a point where you can release a last pre-1.13 build that should fix chests backwards to around 1.8/1.9 (I forget when Inventory#getLocation was added).

>* Breaking changes: we have to consider each of them separately, and separately from the rest of the changes that are just clear improvements.

I believe the only breaking change is to config values that accept material IDs, and that could be fixed with a converter, I just don't really want to write one. A quick search hasn't yielded anything that fits the need, but I'm sure someone's put one out there somewhere.
Beyond that, the 1.13 update really should be done without backwards compatibility, depending on the Spigot's compatibility layer can only do so much. I imagine it'll get worse in future updates. Minecraft and Spigot have been Java 8 since 1.12, there just wasn't a good reason to break backwards compatibility with older servers when it wasn't required.

nitnelave commented:
So, what we can do is the following:

- A series of patches addressing as much as possible in the old environment. That means Java 6, no BlockData, no breaking changes.
- A bump in major version, with Java 8 conversion, BlockData, Material ID changes, dropping Factions support (we'll have to see about that), etc.

For the configuration, I'd really like a silent converter to new style (i.e. if the old field is present, convert it to the new field and drop the old). It doesn't even have to be efficient, you can iterate through all the materials for each value until you find the matching one, it's just a one-time conversion process anyway.

Jikoo committed:
Implement CreeperMultiblock for some blocks

* Fixed double chests/other containers
  * Shulkers will eventually need to be revisited - drop should drop full shulker, not contents separately.
* Reduced redundant explosion replacement entries for "dependent" multiblocks

(cherry picked from commit 2909731b8f7c123fe057c05102ad90f4b9bada0b)

Jikoo commented:
Okay, I fiddled things around a bit, currently should be offering a much nicer diff for pre-1.13 changes.

* Deleted PlayerHeads, it's unused. Is Lockette used at all? I saw mention of it, but no actual usage.
* Updated some block types to use the new Multiblock. Didn't touch pistons or beds, I don't remember them being problematic.
* Reduced redundant code (or lack thereof) in CreeperExplosion's methods for checking exploded blocks, which was the source of the duplication issues for some multiblocks
* Changed plugin.yml to just use the maven version, why update more places than you have to :p
* Did not remove block types that were not necessary due to how BlockStates work

Jikoo committed:
Convert old material ID lists

Jikoo committed:
Java 8/code cleanup

* Lovely lambdas all over
* Type inference is wonderful
* Replaced all Random usage with ThreadLocalRandom
* Restricted member access where possible and not public API

Jikoo commented:
Okay, we're about where I was when I opened the one massive PR. TODO list is updated.

Found a bug with advanced option `drop-destroyed-blocks: false` that appears to have been present [since 2013](https://github.com/nitnelave/CreeperHeal/commit/e53bf58d577607724c9c502487104e0a03b798b3#diff-bbe1d79e8082a86849d6b8c5c3d31b69L123) - drops are always done for whatever block happens to exist in the spot being healed, even if the block is not being overwritten. I'll open a separate PR for that, it should be fixed  before 1.13 and I don't want to have to cherry pick each commit again.

Jikoo committed:
Correctly drop own BlockState drops when breaking

Jukebox support is very hacky, but Spigot 1.13.1 appears to not support removing record correctly. Jukeboxes now effectively have a 1-slot inventory. Bukkit only supports retrieving material, changes do not appear to persist.

Jikoo committed:
Fix Bukkit compatibility layer breaking ID lookup

nitnelave commented:
Thanks for your work! But again, small PRs make happy maintainers :D

Jikoo commented:
>But again, small PRs make happy maintainers :D

1.13 and Java 8 is never going to be small. I've broken it up a little in separate commits, but that's about the best I can do.

I'm going to open a separate pull for pre-1.13 due to how interdependent the bug fixes are with my changes to how certain blockstates are handled, but that's it. I've already spent way too long reconsidering the original diff when splitting it up, I'm not doing it all again over even more content. Completely understand if you don't want to read it all, it's a lot. At least this way it's out there if someone else wants to take a crack at it.

nitnelave commented:
> > But again, small PRs make happy maintainers :D
> 
> 1.13 and Java 8 is never going to be small. I've broken it up a little in separate commits, but that's about the best I can do.

A small PR doesn't necessarily mean few lines changed (although that usually helps), but it can also mean single-purpose. In more mature projects, it's common to have a PR reference a single bug, and address that single bug. Bigger features are implemented in several PR; for instance, you could introduce the new data structures and populate them, implement the algorithm with tests, and then integrate with the main data flow: the first 2 PRs do not affect the control flow of the program and are easily tested on their own, and the last one is usually trivial.

> I'm going to open a separate pull for pre-1.13 due to how interdependent the bug fixes are with my changes to how certain blockstates are handled, but that's it. I've already spent way too long reconsidering the original diff when splitting it up, I'm not doing it all again over even more content. Completely understand if you don't want to read it all, it's a lot. At least this way it's out there if someone else wants to take a crack at it.

However, I understand that this is a minecraft plugin, not an operating system :D We're all volunteers here, I'm just trying to help you improve. I'll try to have a look at the PR when I get the time. Thanks again for your efforts :)

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/PluginHandler.java:
```diff
@@ -18,21 +18,22 @@
         if (lwcp != null)
             lwc = ((LWCPlugin) lwcp).getLWC();
+        else
+            lwc = null;
```
This is not needed, static variables are initialized at the beginning of the execution, with null.

Jikoo replied:
Fair. My compiler complains about the initialization because it's done in and outside of the static block, but that's easy to remove.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/PluginHandler.java:
```diff
@@ -18,21 +18,22 @@
         if (detectPlugin("MobArena") != null)
             maHandler = new MobArenaHandler();
+        else
+            maHandler = null;
```
Same here.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/PluginHandler.java:
```diff
@@ -18,21 +18,22 @@
-    private static LWC lwc = null;
-
-    private static final boolean spout;
+    private static final MobArenaHandler maHandler;
```
Can we try to minimize the diffs? Removing `= null` doesn't achieve much here.

nitnelave commented on pom.xml:
```diff
@@ -40,14 +40,20 @@
         <artifactId>maven-compiler-plugin</artifactId>
         <version>3.1</version>
         <configuration>
-          <source>1.6</source>
```
There's no reason to drop Java 6 support right now, since other servers might not be all the way up to date. We can code in Java 6, unless there's a compelling reason to use Java 8.

Jikoo replied:
There's a commit for Java 8 after the 1.13 commit. 1.12+ use Java 8. Unless you opt to use some sort of [Material replacement](https://www.spigotmc.org/threads/1-8-to-1-13-itemstack-material-version-support.329630/) instead of actually updating to 1.13, there's no point in not using the features available in Java 8.

Jikoo replied:
Going to mark all the other Java 8 notations as resolved, because they all depend on this decision.

nitnelave replied:
Let's do an update to 1.12 and fix all the bugs, and then we'll do a breaking change one.

nitnelave replied:
And for 1.12 your can keep Java 6.

Miclebrick replied:
1.12 itself requires Java 8

nitnelave replied:
Oh, okay. I'd still like to do the Java 8 switch in another PR, for clarity.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -24,7 +29,7 @@
     /*
      * The list of burnt blocks waiting to be replaced.
      */
-    private static List<CreeperBurntBlock> burntList = new LinkedList<CreeperBurntBlock>();
```
Java 8 change.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -39,27 +44,13 @@
     public static void init()
     {
         if (CreeperConfig.getInt(CfgVal.WAIT_BEFORE_BURN_AGAIN) > 0)
-            recentlyBurnt = new HashMap<Location, Date>();
```
Java 8 change.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -39,27 +44,13 @@
-                cleanUp();
-            }
-        }, 300, 7200);
+        Bukkit.getScheduler().scheduleSyncRepeatingTask(CreeperHeal.getInstance(), BurntBlockManager::cleanUp, 300, 7200);
```
Java 8 change.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -39,27 +44,13 @@
-                replaceBurnt();
-            }
-        }, 0, 20);
+        Bukkit.getScheduler().runTaskTimer(CreeperHeal.getInstance(), BurntBlockManager::replaceBurnt, 0, 20);
```
Java 8

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -73,42 +64,42 @@ public static void forceReplaceBurnt(WorldConfig worldConfig)
     {
         World world = Bukkit.getServer().getWorld(worldConfig.getName());
 
-        Iterator<CreeperBurntBlock> iter = burntList.iterator();
```
Java 8 change. You can keep some cosmetic changes (variable names and such), but let's keep it Java 6 for now.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -73,42 +64,42 @@ public static void forceReplaceBurnt(WorldConfig worldConfig)
      */
     public static void forceReplaceBurnt()
     {
-        Iterator<CreeperBurntBlock> iter = burntList.iterator();
```
Same here.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -118,8 +109,8 @@ public static void forceReplaceBurnt()
-        int d = CreeperConfig.getInt(CfgVal.WAIT_BEFORE_BURN_AGAIN);
-        Date time = new Date(new Date().getTime() + 1000 * d);
+        int waitBeforeBurn = CreeperConfig.getInt(CfgVal.WAIT_BEFORE_BURN_AGAIN);
+        Date time = new Date(new Date().getTime() + 1000 * waitBeforeBurn);
```
We probably should have a function for those 2 lines.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/BurntBlockManager.java:
```diff
@@ -223,15 +214,8 @@ private static void cleanUp()
             fireIndex.clean();
         if (CreeperConfig.getInt(CfgVal.WAIT_BEFORE_BURN_AGAIN) > 0)
         {
-            Iterator<Location> iter = recentlyBurnt.keySet().iterator();
```
Java 8 change

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperArmorStand.java:
```diff
@@ -19,7 +19,7 @@
 public class CreeperArmorStand implements Replaceable
 {
     private final ArmorStand stand;
-    private ArrayList<ItemStack> contents = new ArrayList<ItemStack>();
```
Java 8. You can keep the `final`

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBanner.java:
```diff
@@ -1,16 +1,19 @@
 /**
  * Banner implementation of CreeperBlock.
- * 
- * @author drexplosionpd
```
Add yourself to the authors, don't remove the previous guy.

Jikoo replied:
This was actually a [deletion](https://github.com/nitnelave/CreeperHeal/pull/60/commits/42787f31159949b2eaec5bce05e59abe03c3e079#diff-8a8069a7b97a978f01aca8ac4686c101) and [implementation](https://github.com/nitnelave/CreeperHeal/pull/60/commits/f3b468bd793ea17c8393ccde13effe27c8dfc544#diff-8a8069a7b97a978f01aca8ac4686c101). It's not required to special case banners to restore them, which was why this was deleted, that's all it does. However, the bug I found later requires banners to be special cased for dropping the correct item, so I added my own implementation. Looks bad considering the diffs of all the commits as a whole, but that's how we ended up where we are. Can re-add, no problem.

nitnelave replied:
I'd like you to keep the previous author, if it's okay :)

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBanner.java:
```diff
@@ -20,20 +23,21 @@ protected CreeperBanner(Banner banner)
+        if (forced || CreeperConfig.shouldDrop())
+        {
+            ItemStack itemStack = new ItemStack(blockState.getType());
+            BannerMeta bannerMeta = ((BannerMeta) Bukkit.getItemFactory().getItemMeta(blockState.getType()));
```
Is that a 1.13 change?

Jikoo replied:
Nope. I have yet to go back and reconsider the drop fixes for 1.11, I'll be making a separate PR for that when I have time. For now, the fix is at least in 1.13.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java:
```diff
@@ -1,90 +1,62 @@
 /**
- * Bed implementation of CreeperBlock.
- * 
- * @author nitnelave
```
Same here, add yourself to the authors.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java:
```diff
@@ -1,90 +1,62 @@
+
+        BlockData blockData = blockState.getBlockData();
+        if (!(blockData instanceof Bed))
+            throw new IllegalArgumentException("Invalid BlockData: " + blockData.getClass().getName());
```
Are you sure that's not supposed to happen? I.e. there are no "normal" cases where this could throw an exception? Users don't like exceptions.

Jikoo replied:
Since we [check](https://github.com/nitnelave/CreeperHeal/pull/60/commits/42787f31159949b2eaec5bce05e59abe03c3e079#diff-bbe1d79e8082a86849d6b8c5c3d31b69R92) if it's a Bed prior to creation, yes. Someone's gotta be doing something exceptionally weird.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java:
```diff
@@ -1,90 +1,62 @@
+
+        // Ensure materials match.
+        if (head.getType() != foot.getType())
+            return;
```
If they don't match, what do you do? You leave the object half-initialized? If it's an unrecoverable failure that is not supposed to happen, throw an exception, otherwise handle that case anyway.

Jikoo replied:
No, I opted not to add neighbors that don't exist. If the bed is malformed, we consider it half-by-half and restore it individually.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java:
```diff
@@ -1,90 +1,62 @@
+        // Ensure bed halves are correct.
+        if (!(headData instanceof Bed) || !(footData instanceof Bed)
+                || ((Bed) headData).getPart() == ((Bed) footData).getPart())
+            return;
```
Same here.

Jikoo replied:
Abandoned bed half. The BlockState is just the initial block, no dependent blocks as additional parts of the multiblock.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBisected.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+
+/**
+ * Bisected implementation of CreeperMultiblock.
```
Can you add a bit of text about what is a "Bisected"?

Jikoo replied:
Sure. I figured most people looking to work on CH are probably somewhat familiar with [Bukkit's API](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/block/data/Bisected.html). I do note that Spigot's description does not include that stairs and slabs are also bisected while not being 2-block arrangements.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBisected.java:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        if (top.getType() != bottom.getType())
+            return;
```
Same here about handling error cases.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBrick.java:
```diff
@@ -8,12 +9,12 @@
 {
 
-    protected CreeperBrick(BlockState blockState)
+    CreeperBrick(BlockState blockState)
```
Why do you remove `protected`? This class is supposed to only be constructed by the factory in CreeperBlock.

Jikoo replied:
Package-private is more restricted access than protected. Probably should have made a separate commit for restricting member access, I know I did quite a bit of it on non-public things. Sorry about that.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperBrick.java:
```diff
@@ -8,12 +9,12 @@
 
-        if (CreeperConfig.getBool(CfgVal.CRACK_DESTROYED_BRICKS) && getRawData() == (byte) 0)
-            blockState.setRawData((byte) 2);
+        if (CreeperConfig.getBool(CfgVal.CRACK_DESTROYED_BRICKS))
```
I'm not sure this does what we want. There are several types of bricks: mossy, chiseled, plain and cracked. We only want to transform a plain into a cracked.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * InventoryHolder implementation of CreeperMultiblock.
```
A bit more documentation would be nice.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+
+        Inventory inv = blockState.getInventory();
+        if (!(inv instanceof DoubleChestInventory))
+            return;
```
Rather than putting a return, in the case of a constructor, I'd rather put the rest of the function inside the `if`. It strikes me as clearer.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+        World world = blockState.getWorld();
+        if (CreeperConfig.getWorld(world).getBool(WCfgVal.DROP_CHEST_CONTENTS))
+        {
+            Location location = blockState.getLocation().add(0.5, 0.5, 0.5);
```
Nice touch :)

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+            {
+                if (!(dependent instanceof Container))
+                    continue;
+                location = dependent.getLocation().add(0.5, 0.5, 0.5);
```
We should extract dropping the contents into a function, to avoid duplicate code.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+            }
+            return true;
+        }
+        return false;
```
You can put the `return false;` at the beginning right after the check.

Jikoo replied:
Can do. I was trying to be consistent with the other implementations.

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+        blockState.update(true, false);
+
+        dependents.stream().map(BlockState::getChunk).distinct().forEach(Chunk::load);
+        dependents.forEach(state ->
```
Let's try to keep it Java 6

nitnelave commented on src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java:
```diff
@@ -0,0 +1,140 @@
+    @Override
+    public void update()
+    {
+        blockState.getChunk().load();
```
Can you comment on the chunk load? Why is it necessary? Why do you do it only there?

Jikoo replied:
I believe it's completely unnecessary, but that's [how it's been done](https://github.com/nitnelave/CreeperHeal/blob/master/src/main/java/com/nitnelave/CreeperHeal/block/CreeperBlock.java#L246).

nitnelave replied:
It's kinda hard to check, so let's keep it for now.

Miclebrick replied:
Setting and getting blocks via the bukkit API always gets the chunk first, which always loads it if it's not loaded.

Miclebrick replied:
Setting and getting blocks via the bukkit API always gets the chunk first, which always loads it if it's not loaded.

Miclebrick replied:
(thats also a big part of why it's so slow for mass changes, that and relighting and physics)

nitnelave replied:
Then feel free to remove it!

nitnelave reviewed (changes requested):
I didn't have time to review everything, but that should already give you something to do :)

Miclebrick commented:
MC itself actually requires Java 8 now

Jikoo commented:
Thanks for your comments! I've been over all of them and marked duplicates as resolved for readability. Any issues now should be acknowledged or explained, let me know how you want me to proceed.

Jikoo commented:
Actually, were you not aware that 1.8 is required for MC now? In that case, I apologize for my previous attitude, which was getting to be brusque at best. We've been looking at this through completely different eyes for a while. To me, this PR really was for only 1 upgrade which turned into 2 bugfixes and an upgrade (still need to backport the drop fixes, will get to that). You've been seeing it as 2 separate MAJOR changes along with some minor stuff this whole time.

nitnelave commented:
No, I wasn't aware, I haven't really touched any plugin for at least 5 years now. I'd still like to keep Java 6 for 1.12, and then we can update in the next PR.

Jikoo commented:
> I'd still like to keep Java 6 for 1.12, and then we can update in the next PR.

I was planning on keeping Java 6 till 1.13, that was the idea behind 8dca5d7 preceding the other commits. I'll open a separate PR for that, we can discuss it, and then I'll open another PR cherry picking relevant sections of f3b468b in.

When that's out of the way we can get back into this mess, which hopefully I'll have somewhat addressed.

>I think an external IM chat platform, such as Slack or Discord, might be more convenient for this 😛

I work 10+ hours a day, my limited availability probably actually make this easier. No matter how you slice it I end up with a lot of reading to which I get to compose a bulky reply. Here at least we can discuss individual diffs as separate threads, IM would make responding to something specific hours later a nightmare for readability.

Miclebrick commented:
I'm not sure I see the purpose in not updating to Java 6. It's just a change in version number in the build script.

nitnelave commented:
@Miclebrick The point is to separate the changes to make sure you don't break anything. By moving in small steps, you can make sure you move safely. Ideally, we would have plenty of tests to make sure that you don't break anything, but that would require a major reorganization in the plugin :)

nitnelave commented:
Yes, I'm all for updating to Java 8 if it's required by Minecraft. I just want to give the latest bugfixes for those who might still be in Java 6 and don't want to update, and then do the migration.

nitnelave commented:
> I would like to know when a beta version will be released... I, as well as many others, am in dire need of having Creeper Heal installed. It might be interesting to set up a Jenkins server.

Hi! We're still working on this PR, but keep in mind that we're doing that in our free time while juggling work and family life on the side. We'll release a new version when it's ready, but if you're really pressed for time, you can always grab the latest commit from the PR and build the plugin yourself.

VMBindraban commented:
Maybe a radical idea but is it an idea to make separate versions for 1.13 and 1.12 (and below)?
I have seen this for work for multiple plugins.

Jikoo commented:
Not a fan of sacrificing modern potential for ancient versions. I'm very much aware of how hypocritical that sounds coming from the guy who reintroduced backwards compatibility to OpenInv, but it's not worth the hassle of making sure everything works properly on outdated versions.

Considering that no one else has stepped up to fix CH's multiple issues over the years they've been present for, you're probably going to be hard pressed to find anyone willing to maintain multiple versions of it.

nitnelave commented:
Okay, since the other PR has been accepted, we can start working on this one again :)

Jikoo commented:
>we can start working on this one again

I wish ;-; Need to cherry pick f3b468b to 1.11, then fix armor stands not regenerating with armor.

Jikoo committed:
Correctly drop own BlockState drops when breaking

Jukebox support is hacky, but Spigot 1.13 appears to not support removing record correctly. Jukeboxes now effectively have a 1-slot inventory. Bukkit only supports retrieving material, changes do not appear to persist. Ordinary methods would work for versions <= 1.12, however, there's no reason to not include forwards compatibility where possible.

(cherry picked from commit f3b468b)

# Conflicts:
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBanner.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBisected.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBlock.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java

Jikoo committed:
Properly restore ArmorStands' gear

Jikoo committed:
Spawn in correct location

Makes restore look a lot less weird

Jikoo committed:
Rename helper method for clarity

Jikoo committed:
Merge branch 'bugfix-correct-drops'

# Conflicts:
#	pom.xml
#	src/main/java/com/nitnelave/CreeperHeal/PluginHandler.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBanner.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBed.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBlock.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperBrick.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperButton.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperContainer.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperDoor.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperExplosion.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperFlower.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperJukebox.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperMultiblock.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperPiston.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperShulkerBox.java
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperSign.java
#	src/main/resources/plugin.yml

Jikoo committed:
Merge branch 'bugfix-restore-armor-stand-equipment'

# Conflicts:
#	src/main/java/com/nitnelave/CreeperHeal/block/CreeperArmorStand.java

Jikoo committed:
Handle off hand slot for armor stands

Jikoo committed:
Update to 1.13.2

Jikoo commented:
I'm sorry, but I am not going to rework this pull into a satisfactory state. I'm only just getting back to CreeperHeal for a couple bugfixes on 1.15, and I do not have the motivation to delve into finalizing everything on a version so old - it's been nearly 2 years since 1.13's release.
